components/auth/LoginForm.js

```diff
@@ -35,137 +49,131 @@ const fields = {
-    if (fields[key] && fields[key].validator) {
-      errors[key] = fields[key].validator(values);
+  loginFieldsKeys.forEach(key => {
+    if (loginFields[key] && loginFields[key].validator) {
+      const error = loginFields[key].validator(values);
+      if (error) {
+        errors[key] = error;
+      }
     }
   });
   return errors;
 };
 
-const keys = Object.keys(fields);
-const errorsPropsTypes = {};
-keys.forEach(field => {
-  errorsPropsTypes[field] = PropTypes.string;
-});
-
-const LoginForm = ({ onSubmit, pending, errors, allowSignUp }) => {
-  let serverErrors = { ...errors };
-
-  const handleModeSwitch = (formApi, value) => {
-    const clearTouch = {};
-    formApi.setValue('signUp', value);
-    Object.keys(fields).forEach(field => {
-      if (!formApi.values[field]) {
-        clearTouch[field] = false;
-      }
-    });
-    formApi.setAllTouched(clearTouch);
-  };
-
-  return (
-    <div>
-      <h1 className="title is-size-5 has-text-centered">
-        <Text id="auth.signIn" />
-      </h1>
-      <Form onSubmit={onSubmit} validateError={errorValidator}>
-        {formApi => {
-          // FIXME(@drapegnik): find more clean way to handle server-side errors
-          if (serverErrors) {
-            Object.keys(fields).forEach(key => {
-              formApi.setError(key, serverErrors[key]);
-            });
-            serverErrors = null;
-          }
-          return (
-            <form onSubmit={formApi.submitForm}>
-              {allowSignUp && (
-                <div className="field">
-                  <div className="control has-text-centered">
-                    <label htmlFor="signUp" className="checkbox">
-                      <Checkbox
-                        id="signUp"
-                        field="signUp"
-                        onChange={handleModeSwitch.bind(null, formApi)}
-                      />
-                      <Text id="auth.noAccount" />
-                    </label>
-                  </div>
-                </div>
-              )}
-              {!allowSignUp && (
-                <div className="signup-not-available">
-                  <Text id="auth.signup-not-available" />
-                </div>
-              )}
-              {keys.filter(key => formApi.values.signUp || !fields[key].onlyOnSignUp).map(key => (
+const LoginForm = ({ url, allowSignUp, pending }) => (
+  <div>
+    <h1 className="title is-size-5 has-text-centered">
+      <Text id="auth.signIn" />
+    </h1>
+    <Form
+      initialValues={loginInitialValues}
+      validate={loginValidator}
+      action={actions.signIn}
+      handleSubmit={(dispatchAction, { signUp, ...userData }, { setSubmitting }) => {
+        const {
+          query: { lang, next = `/${lang}/articles` },
+        } = url;
+        dispatchAction({ ...userData }, signUp).then(() => Router.pushRoute(next));
+        setSubmitting(false);
+      }}
+    >
+      {({ values, errors, touched, handleSubmit, isSubmitting }) => (
```

Comment:
I think u can use 'render' prop instead children

Reply:
Could you argue your proposal?